### ISSUE TITLE
realtime chart from apex

### ISSUE THREAD

**Issue by vks2** [lovasoa/SQLpage#404]
Hi!
Any chance to implement this?
https://apexcharts.com/javascript-chart-demos/line-charts/realtime/
thanks much

**Comment by lovasoa** [comment]
Hi ! Yes, that would be possible... Can you tell us more about your use case ?

**Comment by vks2** [comment]
thnx for your short reply. i want to pick a Remote Content Demo and adopt it to make it constatnly fetching new data without re-rendering the whole page. i played with shell refresh in 3 sec,  but it refreshes everything despite of embed  param (as described here https://github.com/lovasoa/SQLpage/pull/175). and here you said that we may use LISTEN/NOTIFY in similar scenario. https://github.com/lovasoa/SQLpage/discussions/312 having a very major postgresql background it seems very interesting to me, but we do not have enough human resources to dig in js-rendering issues (and i haven't found any working code about it). finally i found that pattern, that uses an infinite loop https://github.com/lovasoa/SQLpage/blob/v0.22.0/examples/corporate-conundrum/wait.sql but it's too time consuming to adopt my scenario. Anyways i need to silently refresh page when new data is fetched (ideally every 30 secs) and i have a full control of backend postgres (it's a major datawarehouse).

**Comment by lovasoa** [comment]
I get what you want ! Let's keep this issue open until it's implemented.

In the meantime, you can use the card component with iframe as embed mode, and set use the shell refresh in the embedded page:

https://replit.com/@pimaj62145/SQLPage-live-refresh#index.sql

```sql
SELECT 'card' AS component, 1 as columns;

select 'plot.sql' as embed, 'iframe' as embed_mode, 'height: 400px' as style;
```

and in `plot.sql`

```sql
select 'shell' as component, 1 as refresh;

select 'chart' as component, 'y = f(x)' as title;
select x, y from fakedata;
```

**Comment by vks2** [comment]
I'm sorry, with several datatables as embed components, is there a possibility to use form with auto_submit and only one switch type to get param '1 as refresh' working in the embed part? the idea is to rerender some table automatically, some - not.  cannot get it to work in a rational / documented way. THANKS MUCH @lovasoa ! smth like  

```sql
SELECT 'form' AS component, TRUE   as auto_submit;  
 SELECT 'switch' AS type,  'auto_refresh' AS name,  'Auto refresh' AS label; select '_details.sql?_sqlpage_embed&refresh=$checked' as embed;
```

**Comment by lovasoa** [comment]
Here is a way you could do it

## `index.sql`

```sql
SELECT 'form' AS component,
  TRUE as auto_submit,
  'get' as method;

 SELECT 'switch' AS type, 
   'auto_refresh' AS name,  
   $auto_refresh is not null as checked,
   'Auto refresh' AS label;

SELECT 'card' AS component, 1 as columns;

select
  sqlpage.link('liveplot.sql', json_object(
    'auto_refresh', $auto_refresh
   )) as embed,
  'iframe' as embed_mode,
  'height: 400px' as style;
```

## `liveplot.sql`

```sql
select 'shell' as component,
  $auto_refresh is not null as refresh;

select 'chart' as component, 'y = f(x)' as title;
select 1 as x, random() as y;
```

**Comment by vks2** [comment]
Great! You're like Henry Ford running his Ford.
Btw, that refreshes the plot only once when the switch is toggled, it's ok but do not create a partial constant autorefresh of embed component in a page.

**Comment by vks2** [comment]
that do the thing
```sql
select 'shell' as component,
  $auto_refresh is not null as refresh;
select 
    'html' as component,
    case when  $auto_refresh is not null  then ' <script>setTimeout(() => location.reload(), 5000);</script>' else '' end as html;  
```
cheers!

**Comment by lovasoa** [comment]
I don't understand what you are trying to do with that script. What's the point of reloading every 5 seconds in js if you are already reloading every second using the shell component ?

**Comment by vks2** [comment]
with shell component it reloads only once when the switch is toggled

**Comment by lovasoa** [comment]
No, using the shell component is enough to get it to reload continuously. Each time it reloads with the same parameters, 1 as refresh is set again, and it re-refreshes after 1 second.

https://github.com/user-attachments/assets/2df7adbe-1fb6-4752-bfcc-c9124221ce29

**Comment by vks2** [comment]
https://github.com/user-attachments/assets/f4f28cef-4df1-448d-95a0-cec33b2a2764

**Comment by vks2** [comment]
don't work for me. latest sqlpage, windows, tested in Chrome and MS Edge

**Comment by lovasoa** [comment]
The video is too small, I can't see what you are doing. Can you paste your code here ?

**Comment by vks2** [comment]
10  mb per file, sorry.the code is exatcly as yours here, except for filenames.

**Comment by vks2** [comment]
```sql
SELECT 'form' AS component,
  TRUE as auto_submit,
  'get' as method;

 SELECT 'switch' AS type, 
   'auto_refresh' AS name,  
   $auto_refresh is not null as checked,
   'Auto refresh' AS label;

SELECT 'card' AS component, 1 as columns;

select
  sqlpage.link('_d.sql', json_object(
    'auto_refresh', $auto_refresh
   )) as embed,
  'iframe' as embed_mode,
  'height: 400px' as style;
```

and _d.sql

```sql
select 'shell' as component,
  $auto_refresh is not null as refresh;
-- select 
--     'html' as component,
--     case when  $auto_refresh is not null  then ' <script>setTimeout(() => location.reload(), 5000);</script>' else '' end as html;  
select 'chart' as component, 'y = f(x)' as title;
select 1 as x, random() as y;

**Comment by vks2** [comment]
that feedback indeed has low priority (but still i'm facing the issue), so keep good work. thanks much

**Comment by lovasoa** [comment]
https://github.com/user-attachments/assets/e29fab93-7fe2-4283-a9db-b9963c0cd2ff

I can confirm your code snippet above works fine without uncommenting the javascript.
